scripts/open_url_in_instance.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2019 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+# Copyright 2019 Gart Maddy (NoSuck) <admin@nosuck.org>
+
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Script for faster execution.
+
+This script greatly speeds up execution of qutebrowser commands when a
+qutebrowser instance already exists.  Call it in place of your system's
+qutebrowser executable.
+"""
+
+import hashlib
+import json
+import os
+import os.path
+import socket
+import subprocess
+import sys
+
+version = '1.6.0'
+protocol_version = 1
+
+m = hashlib.md5()
+m.update(os.environ['USER'].encode('utf8'))
+socket_name = 'ipc-' + m.hexdigest()
+
+socket_path = os.path.join(os.environ['XDG_RUNTIME_DIR'], 'qutebrowser', socket_name)
+
+command = {
+    'args' : sys.argv[1:],
+    'target_arg' : 'null',
+    'version' : version,
+    'protocol_version' : protocol_version,
+    'cwd' : os.getcwd()
+}
+
+try:
+    s = socket.socket(socket.AF_UNIX)
+    s.connect(socket_path)
+    s.sendall((json.dumps(command) + '\n').encode('utf8'))
+    s.close()
+except OSError:
+    os.environ['PATH'] = os.pathsep.join(reversed(os.get_exec_path()))
```

Comment:
Oh, I'm a bit surprised that setting `$PATH` right before `subprocess.run` works. Can you please test that it actually does? Judging from the line below, you didn't do so yet :wink:

Reply:
Hehe, sorry.  I had a dangling commit.  It's been updated.

Reply:
I was a bit surprised, as well.  Here is what I found, however:

* [subprocess.run](https://github.com/python/cpython/blob/master/Lib/subprocess.py#L474) `with Popen(*popenargs, **kwargs) as process:`
* [subprocess.Popen](https://github.com/python/cpython/blob/master/Lib/subprocess.py#L1580) `for dir in os.get_exec_path(env))`
* [os.get_env_path](https://github.com/python/cpython/blob/master/Lib/os.py#L623) `env = environ`

It all comes back to `os.environ`.  Barring an addition of `os.set_exec_path`, this seems to be the best solution.

Reply:
Testing confirms this on Arch Linux.